File: Meta/3816.lexicographically-smallest-string-after-deleting-duplicate-characters.py
#
# @lc app=leetcode id=3816 lang=python3
#
# [3816] Lexicographically Smallest String After Deleting Duplicate Characters
#
# https://leetcode.com/problems/lexicographically-smallest-string-after-deleting-duplicate-characters/description/
#
# algorithms
# Hard (21.10%)
# Likes:    45
# Dislikes: 4
# Total Accepted:    6.1K
# Total Submissions: 29.1K
# Testcase Example:  "\"aaccb\""
#
#
# You are given a string s that consists of lowercase English letters.
#
# You can perform the following operation any number of times (possibly
# zero times):
#
# Choose any letter that appears at least twice in the current string s
# and delete any one occurrence.
#
# Return the lexicographically smallest resulting string that can be
# formed this way.
#
# Example 1:
#
# Input: s = "aaccb"
#
# Output: "aacb"
#
# Explanation:
#
# We can form the strings "acb", "aacb", "accb", and "aaccb". "aacb" is
# the lexicographically smallest one.
#
# For example, we can obtain "aacb" by choosing 'c' and deleting its first
# occurrence.
#
# Example 2:
#
# Input: s = "z"
#
# Output: "z"
#
# Explanation:
#
# We cannot perform any operations. The only string we can form is "z".
#
# Constraints:
#
# 1 <= s.length <= 10^5
#
# s contains lowercase English letters only.
#

# @lc code=start
class Solution:
    def lexSmallestAfterDeletion(self, s: str) -> str:
        """
        Interview explanation:
        Delete duplicate letters until each remaining letter is unique once, and
        the resulting subsequence is lexicographically smallest.

        Algorithm:
        - Keep last occurrence of each distinct letter as a "must cover" deadline.
        - Greedily append the smallest letter that still has an occurrence after
          the previous pick and leaves all other missing letters finishable.
        - Repeat until every distinct letter has been chosen once.

        Complexity: O(26 * n) time, O(n) space.
        """
        positions = [[] for _ in range(26)]
        for index, ch in enumerate(s):
            positions[ord(ch) - ord("a")].append(index)

        last = [-1] * 26
        missing = 0
        for ch in range(26):
            if positions[ch]:
                last[ch] = positions[ch][-1]
                missing |= 1 << ch

        ptr = [0] * 26
        start = 0
        answer = []

        while missing:
            for ch in range(26):
                pos_list = positions[ch]
                while ptr[ch] < len(pos_list) and pos_list[ptr[ch]] < start:
                    ptr[ch] += 1

                if ptr[ch] == len(pos_list):
                    continue

                index = pos_list[ptr[ch]]
                new_missing = missing & ~(1 << ch)
                if self._can_finish_after(index, new_missing, last):
                    answer.append(chr(ord("a") + ch))
                    start = index + 1
                    missing = new_missing
                    break

        return "".join(answer)

    def _can_finish_after(self, index: int, missing: int, last: list[int]) -> bool:
        for ch in range(26):
            if missing >> ch & 1 and last[ch] <= index:
                return False
        return True
# @lc code=end

File: pt-1/3816.lexicographically-smallest-string-after-deleting-duplicate-characters.go
package leetcode

//
// @lc app=leetcode id=3816 lang=golang
//
// [3816] Lexicographically Smallest String After Deleting Duplicate Characters
//
// Notes
// This is a greedy subsequence construction with one required copy of every
// distinct character. Try letters from 'a' to 'z' at the current start; choose
// the first occurrence that still leaves a future occurrence of every missing
// character. Position lists plus moving pointers make each candidate lookup
// efficient. Time: O(26^2 + n). Space: O(n).
//
// @lc code=start

func LexSmallestAfterDeletion3816(s string) string {
	positions := make([][]int, 26)
	for index, ch := range []byte(s) {
		positions[int(ch-'a')] = append(positions[int(ch-'a')], index)
	}

	last := make([]int, 26)
	for i := range last {
		last[i] = -1
	}
	missing := 0
	for ch := 0; ch < 26; ch++ {
		if len(positions[ch]) > 0 {
			last[ch] = positions[ch][len(positions[ch])-1]
			missing |= 1 << ch
		}
	}

	ptr := make([]int, 26)
	start := 0
	answer := make([]byte, 0, 26)

	for missing != 0 {
		for ch := 0; ch < 26; ch++ {
			posList := positions[ch]
			for ptr[ch] < len(posList) && posList[ptr[ch]] < start {
				ptr[ch]++
			}
			if ptr[ch] == len(posList) {
				continue
			}

			index := posList[ptr[ch]]
			newMissing := missing &^ (1 << ch)
			if canFinishAfter3816(index, newMissing, last) {
				answer = append(answer, byte('a'+ch))
				start = index + 1
				missing = newMissing
				break
			}
		}
	}

	return string(answer)
}

func canFinishAfter3816(index int, missing int, last []int) bool {
	for ch := 0; ch < 26; ch++ {
		if ((missing>>ch)&1) == 1 && last[ch] <= index {
			return false
		}
	}
	return true
}

// @lc code=end
